src/main/host/descriptor/descriptor_table.rs

```diff
@@ -2,49 +2,53 @@ use super::CompatDescriptor;
 use crate::cshadow;
 use crate::utility::notnull::*;
 use log::*;
+use std::collections::{BTreeSet, HashMap};
 use std::convert::TryInto;
 
 /// Table of (file) descriptors. Typically owned by a Process.
 pub struct DescriptorTable {
-    descriptors: std::collections::HashMap<u32, CompatDescriptor>,
-    available_indices: Vec<u32>,
-    index_counter: u32,
+    descriptors: HashMap<u32, CompatDescriptor>,
+
+    // Indices less than `next_index` known to be available.
+    available_indices: BTreeSet<u32>,
+
+    // Lowest index not in `available_indices` that *might* be available. We still need to verify
+    // availability in `descriptors`, though.
+    next_index: u32,
 }
 
 impl DescriptorTable {
     pub fn new() -> Self {
         DescriptorTable {
-            descriptors: std::collections::HashMap::new(),
-            available_indices: Vec::new(),
-            index_counter: 0,
+            descriptors: HashMap::new(),
+            available_indices: BTreeSet::new(),
+            next_index: 0,
         }
     }
 
     /// Add the descriptor at an unused index, and return the index.
     pub fn add(&mut self, descriptor: CompatDescriptor) -> u32 {
         let mut descriptor = descriptor;
 
-        let idx = loop {
-            let idx = match self.available_indices.pop() {
-                Some(i) => {
-                    trace!("Reusing index {}", i);
-                    i
-                }
-                None => {
-                    let i = self.index_counter;
-                    self.index_counter += 1;
-                    trace!("New index {}", i);
-                    i
-                }
-            };
-            if self.descriptors.contains_key(&idx) {
-                // Already in use. This can happen after having called `set`,
-                // since it doesn't update `index_counter` or
-                // `available_indices`.  Assuming this is rare, cheapest
-                // solution in memory and CPU is to just loop here.
-                continue;
+        let idx = if let Some(idx) = self.available_indices.iter().next() {
+            // Un-borrow from `available_indices`.
+            let idx = *idx;
+            // Take from `available_indices`
+            trace!("Reusing available index {}", idx);
+            self.available_indices.remove(&idx);
+            idx
```

Comment:
Ah, it's too bad `BTreeSet::pop_first()` isn't stable yet.

Reply:
Yeah :/. I suppose we could backport it like I did for [stream_len](https://github.com/shadow/shadow/blob/dev/src/main/utility/stream_len.rs), but doesn't seem worth it for just this.